include/rmm/mr/device/detail/arena.hpp

```diff
@@ -0,0 +1,415 @@
+   *
+   * @throws `std::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size in bytes of the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* allocate(std::size_t bytes)
+  {
+    lock_guard lock(mtx_);
+    auto const b     = get_block(bytes);
+    auto const alloc = allocate_from_block(b, bytes);
+    return alloc.pointer;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`, and possibly return superblocks to upstream.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   * @return true if the allocation is found, false otherwise
+   */
+  bool deallocate(void* p, std::size_t bytes, cudaStream_t stream)
+  {
+    lock_guard lock(mtx_);
+    bool found = do_deallocate(p, bytes);
+    shrink_arena(stream);
+    return found;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`, keeping all free superblocks.
+   *
+   * This is done when deallocating across arenas. Since we don't have access to the CUDA stream
+   * associated with this arena, it's not safe to return superblocks.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @return true if the allocation is found, false otherwise
+   */
+  bool deallocate(void* p, std::size_t bytes)
+  {
+    lock_guard lock(mtx_);
+    return do_deallocate(p, bytes);
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @return bool if the allocation is found
+   */
+  bool do_deallocate(void* p, std::size_t bytes)
+  {
+    auto b = free_block(p, bytes);
+    coalesce_block(b);
+    return b.is_valid();
+  }
+
+  /**
+   * @brief Does an arena handle blocks of this size?
+   *
+   * @param size The size in bytes to check
+   * @return true if blocks of this size are handled by an arena
+   */
+  static bool handles_size(std::size_t size) { return size <= maximum_allocation_size; }
+
+ private:
+  /// The maximum allocation size handled by arenas.
+  static constexpr std::size_t maximum_allocation_size = block::superblock_size / 2;
+
+  using lock_guard = std::lock_guard<std::mutex>;
+
+  /**
+   * @brief Get an available memory block of at least `size` bytes.
+   *
+   * @param size The number of bytes to allocate
+   * @return block A block of memory of at least `size` bytes
+   */
+  block get_block(std::size_t size)
+  {
+    // Find the first-fit free block.
```

Comment:
Can you document why first fit vs. best fit?

Reply:
Done.